This Jupyter Notebook is auxillary to the following assessment in this week.  To complete this assessment, answer the 7 questions outlined in this document and use the output from your python cells as answers.

Your goal of this assignment is to fit regression models and interpret their paramters.

Run the following cell to initialize your environment and begin the assessment.

In [1]:
import numpy as np
import pandas as pd  
import seaborn as sns 
import statsmodels.api as sm
import matplotlib.pyplot as plt 
from sklearn.datasets import load_diabetes # loading libraries

In [6]:
data = load_diabetes() 
df = pd.DataFrame(data = data.data, columns = data.feature_names) # create a pandas dataframe

In [3]:
df.head() # dataframe preview

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  
0 -0.002592  0.019907 -0.017646  
1 -0.039493 -0.068332 -0.092204  
2 -0.002592  0.002861 -0.025930  
3  0.034309  0.022688 -0.009362  
4 -0.002592 -0.031988 -0.046641

In [5]:
df.keys() # column names

Index(['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6'], dtype='object')

Now that your notebook is ready, begin answering the questions below.

### Questions 1-3

The first three questions utilize the diabetes dataset from week 1. 

Uncomment and run the following code to fit a linear regression model and output the model summary:

In [7]:
linear_regression_model = sm.OLS.from_formula('s6 ~ age + sex + bmi', df) # define a linear regression model with age, sex and bmi as explanatory variables
result = linear_regression_model.fit() # estimate model parameters
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     s6   R-squared:                       0.225
Model:                            OLS   Adj. R-squared:                  0.220
Method:                 Least Squares   F-statistic:                     42.39
Date:                Fri, 01 Nov 2024   Prob (F-statistic):           4.57e-24
Time:                        19:17:39   Log-Likelihood:                 775.34
No. Observations:                 442   AIC:                            -1543.
Df Residuals:                     438   BIC:                            -1526.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1.006e-17      0.002   5.03e-15      1.000      -0.004       0.004
age            0.2149      0.043      4.956      0.000       0.130       0.300
sex            0.1411      0.043      3.298      0.001       0.057       0.225
bmi            0.3365      0.043      7.847      0.000       0.252       0.421
==============================================================================
Omnibus:                        4.733   Durbin-Watson:                   1.746
Prob(Omnibus):                  0.094   Jarque-Bera (JB):                4.495
Skew:                           0.229   Prob(JB):                        0.106
Kurtosis:                       3.187   Cond. No.                         23.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Question: What is the value of the coefficient for predictor __bmi__?

The coefficient for the BMI variable is 0.3365.

#### Question 2:  Are the predictors for this model statistically significant, yes or no? (Hint: What are their p-values?)

Considering an significance level of **0.005**, we can say that all the predictors are statistically significant (p-values < $\alpha$)

In [9]:
linear_regression_model = sm.OLS.from_formula('s6 ~ age + sex + bmi + bp', data = df) # adding the bp variable as a predictor
result = linear_regression_model.fit() # estimate model parameters
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     s6   R-squared:                       0.256
Model:                            OLS   Adj. R-squared:                  0.249
Method:                 Least Squares   F-statistic:                     37.54
Date:                Fri, 01 Nov 2024   Prob (F-statistic):           5.35e-27
Time:                        19:22:41   Log-Likelihood:                 784.28
No. Observations:                 442   AIC:                            -1559.
Df Residuals:                     437   BIC:                            -1538.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1.006e-17      0.002   5.13e-15      1.000      -0.004       0.004
age            0.1654      0.044      3.749      0.000       0.079       0.252
sex            0.1068      0.043      2.498      0.013       0.023       0.191
bmi            0.2683      0.045      5.961      0.000       0.180       0.357
bp             0.2031      0.048      4.248      0.000       0.109       0.297
==============================================================================
Omnibus:                        5.285   Durbin-Watson:                   1.766
Prob(Omnibus):                  0.071   Jarque-Bera (JB):                5.091
Skew:                           0.255   Prob(JB):                       0.0784
Kurtosis:                       3.124   Cond. No.                         28.4
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Question 3: What happened to our R-Squared value when we added the predictor __bp__ to our initial model?

It increases from **0.225** to **0.256**.

#### Question 4: What type of model might we use when our target outcome, or dependent variable is continuous?

A *linear regression* model.

### Questions 5-6

The next two questions will involve the NHANES dataset.

In [10]:
df = pd.read_csv('nhanes_2015_2016.csv') # read the csv file
df.head() # data preview

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   
3  83735     2.0     1.0     1.0       2         2        56         3   
4  83736     2.0     1.0     1.0       2         2        42         4   

   DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0       1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1       2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2       1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   
3       1.0       5.0  ...   134.0    68.0  109.8  160.9    42.4    38.5   
4       1.0       4.0  ...   114.0    54.0   55.2  164.9    20.3    37.4   

   BMXARML  BMXARMC  BMXWAIST  HIQ210  
0     43.6     35.9     101.1     2.0  
1     40.0     33.2     107.9     NaN  
2     37.0     31.0     116.5     2.0  
3     37.7     38.3     110.1     2.0  
4     36.0     27.2      80.4     2.0  

[5 rows x 28 columns]

In [11]:
df.keys() # column names

Index(['SEQN', 'ALQ101', 'ALQ110', 'ALQ130', 'SMQ020', 'RIAGENDR', 'RIDAGEYR',
       'RIDRETH1', 'DMDCITZN', 'DMDEDUC2', 'DMDMARTL', 'DMDHHSIZ', 'WTINT2YR',
       'SDMVPSU', 'SDMVSTRA', 'INDFMPIR', 'BPXSY1', 'BPXDI1', 'BPXSY2',
       'BPXDI2', 'BMXWT', 'BMXHT', 'BMXBMI', 'BMXLEG', 'BMXARML', 'BMXARMC',
       'BMXWAIST', 'HIQ210'],
      dtype='object')

### Preprocessing Data

In [12]:
df['RIAGENDRx'] = df.RIAGENDR.replace({1: 'Male', 2: 'Female'}) 
df['smq'] = df.SMQ020.replace({2: 0, 7: np.nan, 9: np.nan})
df['DMDEDUC2x'] = df.DMDEDUC2.replace({1: 'lt9', 2: 'x9_11', 3: 'HS', 4: 'SomeCollege',
                                       5: 'College', 7: np.nan, 9: np.nan}) # recoding categorical variables

In [14]:
log_regression_model = sm.GLM.from_formula('smq ~ RIAGENDRx + RIDAGEYR + DMDEDUC2x', 
                                              family = sm.families.Binomial(), data = df) # fitting a logistic regression model
result = log_regression_model.fit() # estimate model parameters
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:                    smq   No. Observations:                 5463
Model:                            GLM   Df Residuals:                     5456
Model Family:                Binomial   Df Model:                            6
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -3421.3
Date:                Fri, 01 Nov 2024   Deviance:                       6842.6
Time:                        19:45:36   Pearson chi2:                 5.47e+03
No. Iterations:                     4   Pseudo R-squ. (CS):             0.1019
Covariance Type:            nonrobust                                         
============================================================================================
                               coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept                   -2.3330      0.111    -21.108      0.000      -2.550      -2.116
RIAGENDRx[T.Male]            0.9313      0.058     15.986      0.000       0.817       1.045
DMDEDUC2x[T.HS]              0.9345      0.087     10.761      0.000       0.764       1.105
DMDEDUC2x[T.SomeCollege]     0.8425      0.082     10.338      0.000       0.683       1.002
DMDEDUC2x[T.lt9]             0.2357      0.106      2.230      0.026       0.029       0.443
DMDEDUC2x[T.x9_11]           1.0745      0.103     10.426      0.000       0.872       1.276
RIDAGEYR                     0.0185      0.002     11.061      0.000       0.015       0.022
============================================================================================
"""

#### Question 5: Which of our predictors has the largest coefficient?

The variable `DMDEDUC2x[T.x9_11]`.

#### Question 6: Which values for DMDEDUC2x and RIAGENDRx are represented in our intercept, or what is our reference level?

**Female** and **College** are the reference levels.

#### Question 7: What model should we use when our target outcome, or dependent variable is binary, or only has two outputs, 0 and 1.

A *logistic regression* model.